packages/php-wasm/universal/src/lib/php-request-handler.ts

```diff
@@ -525,6 +525,38 @@ export class PHPRequestHandler {
 			this.#cookieStore.rememberCookiesFromResponseHeaders(
 				response.headers
 			);
+
+			/**
+			 * Safari has a bug that prevents Service Workers from redirecting relative URLs.
+			 * When attempting to redirect to a relative URL, the network request will return an error.
+			 * See the Webkit bug for details: https://bugs.webkit.org/show_bug.cgi?id=282427.
+			 *
+			 * Because PHP and WordPress can redirect to both relative and absolute URLs
+			 * using the `location` header we need to ensure redirects are processed
+			 * correctly by the Service Worker.
+			 *
+			 * As a workaround for Safari Service Workers, we convert the `location` header
+			 * to an absolute URL for all redirect responses (300-399 status codes).
+			 * This resolves the issue because Safari correctly handles redirects
+			 * to absolute URLs provided by the `location` header.
+			 */
+			if (
+				response.httpStatusCode >= 300 &&
+				response.httpStatusCode <= 399 &&
+				response.headers['location']
+			) {
+				response.headers['location'] = response.headers['location'].map(
+					(location) => {
+						if (isAbsoluteUrl(location)) {
+							return location;
+						}
+						return appendPathnameToBaseUrlAndMaintainScope(
+							new URL(this.#ABSOLUTE_URL),
+							location
+						);
```

Comment:
Why is this needed? Can we just use the [`URL(path, baseUrl)` constructor](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL#syntax) instead to preserve the scope in already-scoped base URLs?

Reply:
Like:
```ts
new URL(location, request.url)
```

The constructor handles the base URL properly and appends the location to the last base path component that looks like a directory, so even if the base URL is http://example.com/subdir/test.php, the result would be based upon `http://example.com/subdir/`.

Reply:
The base scope is lost if I use `new URL(location, request.url)`, the base pathname is replaced by the location pathname.

![Screenshot 2024-11-07 at 08 24 08](https://github.com/user-attachments/assets/29c0a823-9f9f-412e-a4fc-0b31fd8c68c1)

I ended up building this "complex" version because it needs to account for different scenarios. 

1. A Plugin or WP makes a redirect to a pathname without scope (e.g. `header('Location: /wp-admin/options-general.php?page=test');`)
This location header won't include a scope and we need to prepend it.
2. The redirect has scope (e.g. a [reload as we do in autologin](https://github.com/WordPress/wordpress-playground/blob/bef1be234a5d11b322292380303e5716db3e5108/packages/playground/wordpress/src/index.ts#L165)).
The location will include a scope and we need to preserve it.

Reply:
I documented this in the code https://github.com/WordPress/wordpress-playground/pull/1978/files#diff-1846181c9ea179d1066838c075e2f452dfc8fc7dbad5b05345dc9f9895f5810eR166-R185

Reply:
> A Plugin or WP makes a redirect to a pathname without scope (e.g. header('Location: /wp-admin/options-general.php?page=test');)
This location header won't include a scope and we need to prepend it.

That wouldn’t work in hosted WordPress, let’s not bother trying to make it work in Playground.